src/azure-cli-core/azure/cli/core/_identity.py

```diff
@@ -0,0 +1,765 @@
+        if matched_with_tenant:
+            cred = matched_with_tenant[0]
+        else:
+            logger.warning("Could not retrieve credential from local cache for service principal %s under tenant %s. "
+                           "Trying credential under tenant %s, assuming that is an app credential.",
+                           sp_id, tenant, matched[0][_SERVICE_PRINCIPAL_TENANT])
+            cred = matched[0]
+        return cred.get(_ACCESS_TOKEN, None), cred.get(_SERVICE_PRINCIPAL_CERT_FILE, None)
+
+    def save_service_principal_cred(self, sp_entry):
+        self._load_cached_creds()
+        matched = [x for x in self._service_principal_creds
+                   if sp_entry[_SERVICE_PRINCIPAL_ID] == x[_SERVICE_PRINCIPAL_ID] and
+                   sp_entry[_SERVICE_PRINCIPAL_TENANT] == x[_SERVICE_PRINCIPAL_TENANT]]
+        state_changed = False
+        if matched:
+            # pylint: disable=line-too-long
+            if (sp_entry.get(_ACCESS_TOKEN, None) != matched[0].get(_ACCESS_TOKEN, None) or
+                    sp_entry.get(_SERVICE_PRINCIPAL_CERT_FILE, None) != matched[0].get(_SERVICE_PRINCIPAL_CERT_FILE,
+                                                                                       None)):
+                self._service_principal_creds.remove(matched[0])
+                self._service_principal_creds.append(sp_entry)
+                state_changed = True
+        else:
+            self._service_principal_creds.append(sp_entry)
+            state_changed = True
+
+        if state_changed:
+            self._persist_cached_creds()
+
+    def remove_cached_creds(self, user_or_sp):
+        self._load_cached_creds()
+        state_changed = False
+
+        # clear service principal creds
+        matched = [x for x in self._service_principal_creds
+                   if x[_SERVICE_PRINCIPAL_ID] == user_or_sp]
+        if matched:
+            state_changed = True
+            self._service_principal_creds = [x for x in self._service_principal_creds
+                                             if x not in matched]
+
+        if state_changed:
+            self._persist_cached_creds()
+
+    def remove_all_cached_creds(self):
+        try:
+            os.remove(self._token_file)
+        except FileNotFoundError:
+            pass
+
+    def _persist_cached_creds(self):
+        persistence = self._build_persistence()
+        from msal_extensions import CrossPlatLock
+        with CrossPlatLock(self._lock_file):
+            persistence.save(json.dumps(self._service_principal_creds))
+
+    def _load_cached_creds(self):
+        persistence = self._build_persistence()
+        from msal_extensions import CrossPlatLock
+        from msal_extensions.persistence import PersistenceNotFound
+        with CrossPlatLock(self._lock_file):
+            try:
+                self._service_principal_creds = json.loads(persistence.load())
+            except PersistenceNotFound:
```

Comment:
Change to `PersistenceNotFound` due to https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/67.